clients/pkg/logentry/stages/logfmt_test.go

```diff
@@ -0,0 +1,294 @@
+package stages
+
+import (
+	"testing"
+	"time"
+
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/assert"
+	"gopkg.in/yaml.v2"
+)
+
+var testLogfmtYamlSingleStageWithoutSource = `
+pipeline_stages:
+- logfmt:
+    mapping:
+      out:  message
+      app:
+      duration:
+      unknown:
+`
+
+var testLogfmtYamlMultiStageWithSource = `
+pipeline_stages:
+- logfmt:
+    mapping:
+      extra:
+- logfmt:
+    mapping:
+      user:
+    source: extra
+`
+
+var testLogfmtLogLine = `
+	time=2012-11-01T22:08:41+00:00 app=loki	level=WARN duration=125 message="this is a log line" extra="user=foo""
+`
+
+func TestPipeline_Logfmt(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		config          string
+		entry           string
+		expectedExtract map[string]interface{}
+	}{
+		"successfully run a pipeline with 1 json stage without source": {
+			testLogfmtYamlSingleStageWithoutSource,
+			testLogfmtLogLine,
+			map[string]interface{}{
+				"out":      "this is a log line",
+				"app":      "loki",
+				"duration": "125",
+			},
+		},
+		"successfully run a pipeline with 2 json stages with source": {
+			testLogfmtYamlMultiStageWithSource,
+			testLogfmtLogLine,
+			map[string]interface{}{
+				"extra": "user=foo",
+				"user":  "foo",
+			},
+		},
+	}
+
+	for testName, testData := range tests {
+		testData := testData
+
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+
+			pl, err := NewPipeline(util_log.Logger, loadConfig(testData.config), nil, prometheus.DefaultRegisterer)
+			assert.NoError(t, err)
+			out := processEntries(pl, newEntry(nil, nil, testData.entry, time.Now()))[0]
+			assert.Equal(t, testData.expectedExtract, out.Extracted)
+		})
+	}
+}
+
+var testLogfmtCfg = `logfmt:
```

Comment:
I'd move this into the test where it's used. That would make it a little easier to understand the test.